src/DominantJuice.sol

```diff
@@ -0,0 +1,396 @@
+    /// @return delegateAllocations Amount to be sent to delegates instead of adding to local balance. Useful for auto-routing funds from a treasury as payment come in.
+    function payParams(JBPayParamsData calldata _data)
+        external
+        virtual
+        override
+        returns (uint256 weight, string memory memo, JBPayDelegateAllocation3_1_1[] memory delegateAllocations)
+    {
+        // The if statements are the only changes from the payParams() template.
+        // This is also not a payable function, so calls with msg.value will revert.
+        if (getBalance() == 0) revert DataSourceNotInitialized();
+        if (_data.amount.value < minimumPledgeAmount) revert AmountIsBelowMinimumPledge(minimumPledgeAmount);
+
+        // Forward the default weight received from the protocol.
+        weight = _data.weight;
+        // Forward the default memo received from the payer.
+        memo = _data.memo;
+        // Add `this` contract as a Pay Delegate so that it receives a `didPay` call. Don't send any funds to the delegate (keep all funds in the treasury).
+        delegateAllocations = new JBPayDelegateAllocation3_1_1[](1);
+        delegateAllocations[0] = JBPayDelegateAllocation3_1_1(this, 0, "");
+
+        return (weight, memo, delegateAllocations);
+    }
+
+    /// @notice Received hook from the payment terminal after a payment.
+    /// @dev Reverts if the calling contract is not one of the project's terminals.
+    /// @dev This example implementation reverts if the payer isn't on the allow list.
+    /// @param _data Standard Juicebox project payment data.
+    /// See https://docs.juicebox.money/dev/api/data-structures/jbdidpaydata3_1_1/.
+    function didPay(JBDidPayData3_1_1 calldata _data) external payable virtual override {
+        // Make sure the caller is a terminal of the project, and that the call is
+        // being made on behalf of an interaction with the correct project.
+        if (
+            msg.value != 0 || !directory.isTerminalOf(projectId, IJBPaymentTerminal(msg.sender))
+                || _data.projectId != projectId
+        ) revert InvalidPaymentEvent(msg.sender, _data.projectId, msg.value);
+
+        if (block.timestamp >= cycleExpiryDate) revert CycleHasExpired();
+
+        // Get payer address and amount paid from JB DidPay data struct
+        address payer = _data.payer;
+        JBTokenAmount memory amount = _data.amount;
+        uint256 paymentAmount = amount.value;
+
+        // Update storage variables
+        totalAmountPledged += paymentAmount;
+        pledgedAmount[payer] = paymentAmount;
+        pledgers.push(payer);
+        orderToPledgerToAmount[pledgeOrder][payer] = paymentAmount;
+        pledgeOrder++;
+    }
+
+    /**
+     * @notice After the funding cycle ends, anyone can call this function to store the cycle results.
+     * @dev
+     */
+    function relayCycleResults() public {
```

Comment:
what is the value of this function? two getter functions could be written `isCycleExpired()` and `isTargetMet()` without the need to update variables on the contract. This just costs gas with no inherent value.

Reply:
Good points! I can refactor to have two getter functions for sure.

Reply:
I guess the one thing though is that a cycle closed event wouldn't be able to be emitted from a getter. Not sure it's too important, but I like to try to write really transparent and communicative contracts (within reason of course).

Reply:
I agree with transparency....every write should have an event. However, if there are no writes, there are no events. The payments and pledges are the events. But it's cool if you want this function for the event. Just no need to also update storage if you don't need to. It's also kind of weird that anyone can call this function.

Reply:
I'll remove it. That's fine. The pledgers are going to have feedback on the JB site either way. And also, when reconfigureCyclesFor is called, there should be an event emitted, so that will provide feedback as well.